kore/src/Kore/Step/AxiomPattern.hs

```diff
@@ -0,0 +1,60 @@
+{-|
+Description : Common functionality for RulePattern and EqualityPattern
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+-}
+module Kore.Step.AxiomPattern
+    ( HasLeftPattern (..)
+    , HasRequiresPredicate (..)
+    , HasRefreshPattern (..)
+    , HasMapVariables (..)
+    ) where
+
+import Kore.Attribute.Pattern.FreeVariables
+    ( FreeVariables (..)
+    )
+import Kore.Internal.Predicate
+    ( Predicate
+    )
+import Kore.Internal.TermLike
+    ( TermLike
+    )
+import Kore.Substitute
+    ( SubstitutionVariable
+    )
+import Kore.Variables.Fresh
+    ( Renaming
+    )
+
+{-| Should be implemented for rules having a clearly identifiable
+left-hand-side pattern used for matching / unifying with the configuration.
+-}
+class HasLeftPattern rule variable where
```

Comment:
I am generally opposed to having `Has...` classes because 1. they don't actually allow flexible behavior and 2. they constrain all instances to have the same underlying representation without giving the benefit of actually having the same data type. We should either refactor the data types so that all the rules have the same (possibly parameterized) representation _or_ we should introduce a new class whose members are every function that currently has a `HasLeftPattern` constraint. The same goes for `HasRequiresPredicate` and `HasMapVariable`.

Reply:
I've created a `UnifyingRule` class to comprise the common unification code used by both `EqualityPattern` and `RulePattern`.  I agree with you that `HasLeftPattern` and `HasRequires` are not very good.

Nevertheless, I think that a class/classes providing 'freeVariables', 'mapVariables', 'refreshVariables' would actually be beneficial, as we already have these functions implemented for multiple types, sometimes with slight changes in the signature.
Having a class/classes to provide these might recognize the existing common functionality.
What do you think?

Reply:
This sounds like a generalization of the `Binding` class used to implement substitution. What if we implement a `Substitute` class that has `freeVariables`, `refreshVariables`, `mapVariables`, and `traverseVariables` (which basically support capture-avoiding substitution), and also `substitute`?

Reply:
Sounds good!  Should we do it as part of this PR, or open a new one?